File: fcxd/src/linux/fcx_system.c
#include "../fcx_system.h"
#include <pwd.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/utsname.h>
#include <unistd.h>

void fcx_system_info(fcx_system_info_t *info) {
  struct utsname utsname;
  if (uname(&utsname) == 0) {
    snprintf(info->os_version, sizeof(info->os_version), "%s %s",
             utsname.sysname, utsname.release);
    snprintf(info->hostname, sizeof(info->hostname), "%s", utsname.nodename);
  } else {
    info->os_version[0] = '\0';
    info->hostname[0] = '\0';
  }

  struct passwd *pw = getpwuid(getuid());
  if (pw != NULL) {
    snprintf(info->username, sizeof(info->username), "%s", pw->pw_name);
    snprintf(info->full_user_name, sizeof(info->full_user_name), "%s",
             pw->pw_gecos != NULL ? pw->pw_gecos : "");
    snprintf(info->home_directory, sizeof(info->home_directory), "%s",
             pw->pw_dir);
  } else {
    info->username[0] = '\0';
    info->full_user_name[0] = '\0';
    info->home_directory[0] = '\0';
  }
}

File: fcxd/src/mac/fcx_system.m
#include "../fcx_system.h"
#include <Foundation/Foundation.h>
#include <stdio.h>

void fcx_system_info(fcx_system_info_t *info) {
  NSProcessInfo *process = [NSProcessInfo processInfo];

  snprintf(info->os_version, sizeof(info->os_version), "%s",
           [[process operatingSystemVersionString] UTF8String]);
  snprintf(info->username, sizeof(info->username), "%s",
           [NSUserName() UTF8String]);
  snprintf(info->full_user_name, sizeof(info->full_user_name), "%s",
           [NSFullUserName() UTF8String]);
  snprintf(info->home_directory, sizeof(info->home_directory), "%s",
           [NSHomeDirectory() UTF8String]);
  snprintf(info->hostname, sizeof(info->hostname), "%s",
           [[process hostName] UTF8String]);
}
